dali/operators/random/rng_base_gpu.cuh

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_RANDOM_RNG_BASE_GPU_CUH_
+#define DALI_OPERATORS_RANDOM_RNG_BASE_GPU_CUH_
+
+#include <utility>
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/kernels/alloc.h"
+#include "dali/operators/random/rng_base_gpu.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/util/batch_rng.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+
+template <>
+struct RNGBaseFields<GPUBackend> {
+  RNGBaseFields<GPUBackend>(int64_t seed, int max_batch_size)
+      : max_blocks_(std::max(max_batch_size, 1024)),
+        randomizer_(seed, block_size_ * max_blocks_) {
+    block_descs_gpu_ =
+        kernels::memory::alloc_unique<BlockDesc>(kernels::AllocType::GPU, max_blocks_);
+    block_descs_cpu_ =
+        kernels::memory::alloc_unique<BlockDesc>(kernels::AllocType::Pinned, max_blocks_);
+  }
+  static constexpr int block_size_ = 256;
+  const int max_blocks_;
+  curand_states randomizer_;
+  kernels::memory::KernelUniquePtr<BlockDesc> block_descs_gpu_;
+  kernels::memory::KernelUniquePtr<BlockDesc> block_descs_cpu_;
+};
+
+namespace {
+
+template <typename Out, typename Dist, bool DefaultDist = false>
+__global__ void RNGKernel(BlockDesc *block_descs, curandState* states, Dist* dists) {
```

Comment:
Nitpick: use consistent alignment of *
```suggestion
__global__ void RNGKernel(BlockDesc *block_descs, curandState *states, Dist *dists) {
```

Reply:
Also: can it ever happen that dists are non-const? Modifying them constitutes a race condition.

Reply:
will do